operator/auth/spire/client.go

```diff
@@ -0,0 +1,175 @@
+		"SPIRE server endpoint.")
+	flags.StringVar(&cfg.SpiffeTrustDomain,
+		"mesh-auth-spiffe-trust-domain",
+		"spiffe.cilium.io",
+		"The trust domain for the SPIFFE identity.")
+}
+
+type Client struct {
+	cfg   ClientConfig
+	entry entryv1.EntryClient
+}
+
+// NewClient creates a new SPIRE client.
+// If the mTLS is not enabled, it returns a noop client.
+func NewClient(cfg ClientConfig) (identity.Provider, error) {
+	if !cfg.AuthMTLSEnabled {
+		return &noopClient{}, nil
+	}
+
+	// This is blocking till the cilium-operator is registered in SPIRE.
+	source, err := workloadapi.NewX509Source(
+		context.Background(),
+		workloadapi.WithClientOptions(
+			workloadapi.WithAddr(fmt.Sprintf("unix://%s", cfg.SpireAgentSocketPath)),
+		),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	tlsConfig := tlsconfig.MTLSClientConfig(source, source, tlsconfig.AuthorizeAny())
+	conn, err := grpc.Dial(cfg.SpireServerAddress, grpc.WithTransportCredentials(credentials.NewTLS(tlsConfig)))
+	if err != nil {
+		return nil, err
+	}
+
+	return &Client{
+		cfg:   cfg,
+		entry: entryv1.NewEntryClient(conn),
+	}, nil
+}
+
+// Upsert creates or updates the SPIFFE ID for the given ID.
+// The SPIFFE ID is in the form of spiffe://<trust-domain>/cilium-id/<id>.
+func (c *Client) Upsert(ctx context.Context, id string) error {
+	entry, err := c.entry.GetEntry(ctx, &entryv1.GetEntryRequest{Id: id})
+	if err != nil && !strings.Contains(err.Error(), notFoundError) {
+		return err
+	}
+
+	desired := []*types.Entry{
+		{
+			SpiffeId: &types.SPIFFEID{
+				TrustDomain: c.cfg.SpiffeTrustDomain,
+				Path:        fmt.Sprintf("/cilium-id/%s", id),
+			},
+			ParentId: &types.SPIFFEID{
+				TrustDomain: c.cfg.SpiffeTrustDomain,
+				Path:        defaultParentID,
+			},
+			Selectors: defaultSelectors,
+		},
+	}
+
+	if entry == nil {
+		_, err = c.entry.BatchCreateEntry(ctx, &entryv1.BatchCreateEntryRequest{Entries: desired})
+		return err
+	}
+
+	_, err = c.entry.BatchUpdateEntry(ctx, &entryv1.BatchUpdateEntryRequest{
+		Entries: desired,
+	})
+	return err
+}
+
+// Delete deletes the SPIFFE ID for the given ID.
+// The SPIFFE ID is in the form of spiffe://<trust-domain>/cilium-id/<id>.
+func (c *Client) Delete(ctx context.Context, id string) error {
+	entries, err := c.entry.ListEntries(ctx, &entryv1.ListEntriesRequest{
+		Filter: &entryv1.ListEntriesRequest_Filter{
+			BySpiffeId: &types.SPIFFEID{
+				TrustDomain: c.cfg.SpiffeTrustDomain,
+				Path:        fmt.Sprintf("/cilium-id/%s", id),
+			},
+			ByParentId: &types.SPIFFEID{
+				TrustDomain: c.cfg.SpiffeTrustDomain,
+				Path:        defaultParentID,
+			},
+			BySelectors: &types.SelectorMatch{
+				Selectors: defaultSelectors,
+				Match:     types.SelectorMatch_MATCH_EXACT,
+			},
+		},
+	})
+	if err != nil {
+		if strings.Contains(err.Error(), notFoundError) {
+			return nil
+		}
+		return err
+	}
+
+	var ids = make([]string, len(entries.Entries))
+	for _, e := range entries.Entries {
+		ids = append(ids, e.Id)
```

Comment:
if we define the length in the make we might assign the value with index already here as that saves the whole codepath of the append()